keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1065,6 +1065,51 @@ Details to consider:
 
 #### CEL Expression Composition
 
+##### Policy Predicates
+
+The match criteria in bindings are not expected to be able to cover all possible
+ways users may want to scope their policies. For example, there is no way to
+match off of kind, only resource. To provide extensibility for the match critiera
+without requiring modifying every validation rule individually, a global predicate
+system is needed. These predicates contain CEL statements that must be satisfied, otherwise
+the policy will be ignored. In order to keep bindings language-agnostic and to support
+singleton policies, the logic should live in the policy definition resource. To enable
+customization per-binding, the CEL statements should have access to the parameter resource.
+
+Here is an example of a policy definition using policy predicates (under the `customConstraints` field):
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "replicalimit-policy.example.com"
+Spec:
+  failurePolicy: Fail
+  paramKind:
+    apiVersion: rules.example.com/v1
+    kind: ReplicaLimit
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+  customConstraints:
+    - expression: 'metadata.kind == "Deployment"'
+    - expression: '!(metadata.namespace in params.excludedNamespaces)'
+  validations:
+    - expression: "object.spec.replicas <= params.maxReplicas"
+      reason: Invalid
+```
+
+For demonstration purposes, we assume `match` has no support for `excludedNamespaces`.
+
+`customConstraints` has the following behaviors:
+
+* Only the request object and parameters are accessible (no referential lookup)
```

Comment:
These predicates are purely syntactic sugar, right? In other words, any logic that is present in these expressions could be encoded directly in the validations, right?

Given that, I'm wondering whether these expressions should be allowed to access params. Reasons to forbid access to params are:
1. Debuggability - I suspect it will be easier to debug an unexpected validation result than to figure out why a request was ignored. Including params here greatly increases the scope of what needs to be understood to see whether / how a request will be evaluated. If these are scoped to only the `AdmissionReview`, it seems like the logic would be much easier to follow. See also above point about matching multiple sets of parameters.
2. Optimizations - avoid fetching parms if the predicate filters them out. I suppose we could just be smart about evaluating the expressions that don't represent params first, but explicitly forbidding them might encourage better patterns.

Reply:
I'm not sure where the line between utility and "syntactic sugar" is drawn, but I can say that this provides real value for policy framework authors. Without this language element, policy framework authors would need to rewrite the contents of users' CEL expressions, which is a fraught exercise.

It also provides a performance benefit: copy/pasting predicate logic into every validation would mean unnecessary re-evaluation of code.

> Given that, I'm wondering whether these expressions should be allowed to access params.

This would make it difficult or impossible for framework authors to polyfill missing pieces of functionality they need. For example, right now there is no way for Gatekeeper to match against `Kind` (only `resource` is supported).

Other uses, such as universal predicates that allow avoid evaluating the same code for multiple bindings (parameter-less predicates as contemplated by you and jpbetz) are another cool use case, but not the motivation for this proposal.

> Debuggability - I suspect it will be easier to debug an unexpected validation result than to figure out why a request was ignored. Including params here greatly increases the scope of what needs to be understood to see whether / how a request will be evaluated. If these are scoped to only the AdmissionReview, it seems like the logic would be much easier to follow. See also above point about matching multiple sets of parameters.

I'm not sure how this greatly increases the scope of what needs to be understood? From the perspective of someone reading the code, is this any more/less comprehensible than prefixing every `validation` with an equivalent predicate?

If anything, this should increase debug-ability relative to the alternative ("any logic that is present in these expressions could be encoded directly in the validations"):

* potential for less repeated code, making resources more grok-able
* ability to log when a request is ignored b/c of a dissatisfied predicate
* framework authors do not need to worry about potential interactions between users' CEL and their own code

> Optimizations - avoid fetching parms if the predicate filters them out. I suppose we could just be smart about evaluating the expressions that don't represent params first, but explicitly forbidding them might encourage better patterns.

What is the concern with fetching parameters? They are already pre-cached by the controller, so there is no network latency concern. If there is a cost to serialization, we can, as you said, always optimize execution order.

I'm not sure we need to be too concerned about people accessing `params` unnecessarily... it seems to me that the default would be to avoid accessing `params` where possible to avoid the extra work of adding a parameter schema.

Reply:
To be clear, I wasn't doubting the value of this feature addition overall. I just wanted to clarify as we discuss the details whether this is increasing the scope of what is possible with the current implementation.

> For example, right now there is no way for Gatekeeper to match against Kind (only resource is supported).

Does matching Kind need to be parameterized by the bindings though? I'd expect it to be constant for most constraint templates.

> I'm not sure how this greatly increases the scope of what needs to be understood?

My concern isn't about someone understanding the CEL expression or code, but rather the cluster configuration. If I write a rule that isn't matching a request, I need to cross check the predicate rules with all the bindings that could match it, and all there parameters. Conversely, if I want to know whether a validation applies to a given request, I need to think through a lot more objects.

> ability to log when a request is ignored b/c of a dissatisfied predicate

I think this gets at my debuggability concerns. I'd expect we might log the outcome (mostly negative, but maybe positive at higher verbosity) of validations for requests that meet the match criteria, but I wouldn't expect us to log every validation that didn't match a given request. I'm assuming that predicates are treated more like match criteria than validations, but maybe that's an incorrect assumption.

Reply:
> Does matching Kind need to be parameterized by the bindings though? I'd expect it to be constant for most constraint templates.

One clarification -- this proposal does not suggest adding additional configuration to the Binding resource, merely access to the parameters resource.

Metadata-based constraint templates can require specifying kind. More generally, without a way to map Gatekeeper's match criteria onto this KEP, this KEP wouldn't be able to be the backend for constraints that have un-mappable criteria . Another example of a match criteria that will probably require custom logic for Gatekeeper: prefix-based namespace matching.

> If I write a rule that isn't matching a request, I need to cross check the predicate rules with all the bindings that could match it, and all there parameters. Conversely, if I want to know whether a validation applies to a given request, I need to think through a lot more objects.

It's not clear from the wording, but I think what you're saying is that "if a violation isn't matching, I need to check more objects if a predicate can access parameters, vs. if a predicate can't".

I don't think this is true. As you pointed out, predicate-like behavior can be approximated (modulo performance improvements and risk of CEL syntax errors) by prefixing violations with the predicate code. Given that, this proposal has no effect on the number of potential resources that would need to be understood to diagnose an error.

If people are concerned about complexity, and don't need to parameterize, they can just not access parameters. The lack of parameter references in the CEL code would be sufficient to know no more resources need to be grokked.

Conversely, if users do need parameterized match criteria, then the code clarity, performance, syntactic risk, and potential options for debugging are all improved via this proposal.

> but I wouldn't expect us to log every validation that didn't match a given request

Agreed, but you could send "tracer requests" that have increased logging to debug why an unexpected result occurred.

> I'm assuming that predicates are treated more like match criteria than validations, but maybe that's an incorrect assumption.

In my mind they are more akin to validations. Match criteria are hardcoded into the API server, which means we can do things like indexing, derive version conversion, from them, etc. Match criteria are also able to be specified in the Binding resource, predicates cannot.

This is closer to a tweaking of the way we use CEL. The current model for triggering a violation is:

`validation fail OR validation fail OR validation fail`

predicates allow us to normalize what would otherwise be a cross-product and write:

`predicate match AND predicate match AND predicate match AND (validation fail OR validation fail OR validation fail)`

Reply:
Based on your last comment, I wonder if what we really need is a policy representation in conjunctive normal form. Something along the lines of:

```
# (expression1 OR expression2) AND (expression3 OR expression4)
all:
  - any:
    - expression1
    - expression2
  - any:
    - expression3
    - expression4
```

But perhaps that's overcomplicating things?

Reply:
That would definitely do the trick for the use cases I'm most concerned about.

The performance optimizations jpbetz@ mentioned would still be possible as well.

I think the key for determining the most suitable arrangement is to figure out what else we want to do: variables? Variables defined iteratively a la [CEL policy templates](https://github.com/google/cel-policy-templates-go/blob/4f68a4a0f51362b1b593f5f7ae791b715345590b/test/testdata/multiple_ranges/template.yaml#L28-L35)?

If any of these could benefit from a separate cluster of CEL expressions that are evaluated in a different order, then the first proposal might make sense. IMO the second proposal (yours) is syntactically cleaner and probably simpler.

Reply:
OTOH having it as a separate field does make for an easy parallel with the webhook predicate KEP:

https://github.com/kubernetes/enhancements/pull/3717/files

Could probably use the same code for both